File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/general/DecimalAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import java.util.List;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.RegexValidator;
import com.truveta.opentoken.attributes.validation.SerializableAttributeValidator;

/**
 * Represents a generic decimal (floating-point) attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * decimal fields. It recognizes "Decimal" as the primary alias for this attribute type.
 * 
 * The attribute performs normalization on input values, trimming whitespace and
 * ensuring the value is a valid decimal number.
 * 
 * The attribute also performs validation on input values, ensuring they match
 * common decimal number formats (with or without decimal points, scientific notation).
 */
public class DecimalAttribute extends BaseAttribute {

    private static final String NAME = "Decimal";
    private static final String[] ALIASES = new String[] { NAME };

    /**
     * Regular expression pattern for validating decimal format.
     * 
     * This regex validates that the input is:
     * - Optional leading whitespace
     * - Optional leading sign (+ or -)
     * - One or more digits, optionally followed by a decimal point and more digits
     * - OR a decimal point followed by one or more digits
     * - Optional exponent notation (e or E followed by optional sign and digits)
     * - Optional trailing whitespace
     * 
     * Valid examples: 123, -45.67, +89.0, .5, 1., 1.5e10, -3.14E-2
     */
    private static final String DECIMAL_REGEX = "^\\s*[+-]?(\\d+(\\.\\d*)?|\\.\\d+)([eE][+-]?\\d+)?\\s*$";

    public DecimalAttribute() {
        super(List.of(new RegexValidator(DECIMAL_REGEX)));
    }

    /**
     * Protected constructor for subclasses to add additional validators.
     * 
     * @param additionalValidators additional validators to apply beyond the base decimal format validation
     */
    protected DecimalAttribute(List<SerializableAttributeValidator> additionalValidators) {
        super(createValidatorList(additionalValidators));
    }

    private static List<SerializableAttributeValidator> createValidatorList(
            List<SerializableAttributeValidator> additionalValidators) {
        java.util.ArrayList<SerializableAttributeValidator> validators = new java.util.ArrayList<>();
        validators.add(new RegexValidator(DECIMAL_REGEX));
        validators.addAll(additionalValidators);
        return validators;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        if (value == null) {
            throw new IllegalArgumentException("Decimal value cannot be null");
        }

        // Trim whitespace and parse as double
        String trimmed = value.trim();

        try {
            double decimalValue = Double.parseDouble(trimmed);
            // Return the decimal value as a string
            return String.valueOf(decimalValue);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid decimal format: " + value);
        }
    }

}

File: lib/python/opentoken/src/main/opentoken/attributes/general/decimal_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List, Optional
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation import RegexValidator
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class DecimalAttribute(BaseAttribute):
    """Represents a generic decimal (floating-point) attribute.

    This class extends BaseAttribute and provides functionality for working with
    decimal fields. It recognizes "Decimal" as a valid alias for this attribute type.

    The attribute performs normalization on input values by trimming whitespace
    and validates that the input is a valid decimal number.

    Supported formats include:
    - Standard decimals: 123, -45.67, +89.0
    - Decimals without leading digit: .5
    - Decimals without trailing digits: 1.
    - Scientific notation: 1.5e10, -3.14E-2
    """

    NAME = "Decimal"
    ALIASES = [NAME]

    # Regular expression pattern for validating decimal format
    # Supports: optional sign, digits with optional decimal, scientific notation
    VALIDATION_PATTERN = r"^\s*[+-]?(\d+\.\d*|\d*\.\d+|\d+)([eE][+-]?\d+)?\s*$"

    def __init__(self, additional_validators: Optional[List[SerializableAttributeValidator]] = None):
        """
        Initialize the DecimalAttribute with optional additional validators.

        Args:
            additional_validators: Optional list of additional validators to apply
        """
        validation_rules: List[SerializableAttributeValidator] = [RegexValidator(self.VALIDATION_PATTERN)]
        if additional_validators:
            validation_rules.extend(additional_validators)
        super().__init__(validation_rules)

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "Decimal"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalize the decimal value by trimming whitespace and parsing.

        Args:
            value: The decimal string to normalize

        Returns:
            str: The normalized decimal value

        Raises:
            ValueError: If the decimal is not valid or empty
        """
        if value is None:
            raise ValueError("Decimal value cannot be null")
            
        value_stripped = value.strip() if value else ""
        
        if not value_stripped:
            raise ValueError("Invalid decimal format: empty or whitespace")

        # Parse and normalize the decimal value
        try:
            decimal_value = float(value_stripped)
            return str(decimal_value)
        except ValueError:
            raise ValueError(f"Invalid decimal format: {value}")
